wp-includes/components/class-wp-service-worker-offline-commenting-component.php

```diff
@@ -0,0 +1,153 @@
+			 *     @type string $revision Revision for the template. This defaults to the template and stylesheet names, with their respective theme versions.
+			 * }
+			 */
+			$offline_error_precache_entry = apply_filters( 'wp_offline_error_precache_entry', $offline_error_precache_entry );
+
+			/**
+			 * Filters what is precached to serve as the internal server error response on the frontend.
+			 *
+			 * The URL returned in this array will be precached by the service worker and served as the response when
+			 * the server returns a 500 internal server error . To prevent this behavior, this value can be filtered
+			 * to return false. When a theme or plugin makes a change to the response, the revision value in the array
+			 * must be incremented to ensure the URL is re-fetched to store in the precache.
+			 *
+			 * @since 0.2
+			 *
+			 * @param array $entry {
+			 *     Server error precache entry.
+			 *
+			 *     @type string $url      URL to page that shows the server error template.
+			 *     @type string $revision Revision for the template. This defaults to the template and stylesheet names, with their respective theme versions.
+			 * }
+			 */
+			$server_error_precache_entry = apply_filters( 'wp_server_error_precache_entry', $server_error_precache_entry );
+
+		} else {
+			$offline_error_precache_entry = array(
+				'url'      => add_query_arg( 'code', 'offline', admin_url( 'admin-ajax.php?action=wp_error_template' ) ), // Upon core merge, this would use admin_url( 'error.php' ).
+				'revision' => PWA_VERSION, // Upon core merge, this should be the core version.
+			);
+			$server_error_precache_entry  = array(
+				'url'      => add_query_arg( 'code', '500', admin_url( 'admin-ajax.php?action=wp_error_template' ) ), // Upon core merge, this would use admin_url( 'error.php' ).
+				'revision' => PWA_VERSION, // Upon core merge, this should be the core version.
+			);
+		}
+
+		$scripts->register(
+			'wp-offline-commenting',
+			array(
+				'src'  => array( $this, 'get_script' ),
+				'deps' => array( 'wp-base-config' ),
+			)
+		);
+
+		if ( $offline_error_precache_entry ) {
+			$scripts->precaching_routes()->register( $offline_error_precache_entry['url'], isset( $offline_error_precache_entry['revision'] ) ? $offline_error_precache_entry['revision'] : null );
+		}
+		if ( $server_error_precache_entry ) {
+			$scripts->precaching_routes()->register( $server_error_precache_entry['url'], isset( $server_error_precache_entry['revision'] ) ? $server_error_precache_entry['revision'] : null );
+		}
+
+		$this->replacements = array(
+			'ERROR_OFFLINE_URL' => wp_service_worker_json_encode( isset( $offline_error_precache_entry['url'] ) ? $offline_error_precache_entry['url'] : null ),
+			'ERROR_500_URL'     => wp_service_worker_json_encode( isset( $server_error_precache_entry['url'] ) ? $server_error_precache_entry['url'] : null ),
+			'ERROR_MESSAGES'    => wp_service_worker_json_encode( wp_service_worker_get_error_messages() ),
```

Comment:
@westonruter Created a separate component -- above you can see that getting these 3 URLs is just a duplication of Routing Component. Perhaps we should move these out of the components to avoid duplication.